File: algorithms/sorting/mergesort.py
from typing import List
'''
Merge Sort: a divide + conquer algorithm. This algorithm takes two subsets of an array and merges them into one sorted array.
    - 
'''

'''
Have we reached the end of any of the arrays?
    No:
        Compare current elements of both arrays 
        Copy smaller element into sorted array
        Move pointer of element containing smaller element
    Yes:
        Copy all remaining elements of non-empty array
'''
def mergeSort(nums: List[int]):
    # Have we reached the end?
    if len(nums) > 1:
        
        # Split the array into two halves.
        mid = len(nums)//2
        left = nums[:mid]
        right = nums[mid:]

        # Sort the two halves
        mergeSort(left)
        mergeSort(right)

        firstptr = secondptr = tmp = 0

        # Until we reach either end of either left or right, pick larger among
        # elements firstptr and secondptr and place them in the correct position at A[p..r]
        while firstptr < len(left) and secondptr < len(right):
            if left[firstptr] < right[secondptr]:
                nums[tmp] = left[firstptr]
                firstptr += 1
            else:
                nums[tmp] = right[secondptr]
                secondptr += 1
            tmp += 1

        # When we run out of elements in either left or right,
        # pick up the remaining elements and put in A[p..r]
        while firstptr < len(left):
            nums[tmp] = left[firstptr]
            firstptr += 1
            tmp += 1

        while secondptr < len(right):
            nums[tmp] = right[secondptr]
            secondptr += 1
            tmp += 1
    
arr = [6,5,7,10,0,-12,1,4]
print("Unsorted: ", arr)
mergeSort(arr)
print("Sorted: ", arr)

File: algorithms/cpp/sorting/mergesort.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <queue>
// Templates
template<typename T>
void print_array(T const(& arr))
{
    for(auto i : arr) {
        std::cout<<i<<' ';
    }
    std::cout<<""<<std::endl;
}

template<typename T>
void newSwap(T *a, T *b)
{
    T t = *a;
    *a = *b;
    *b = t;
}
/*
 * MergeSort is an example of a divide and conquer algorithm. The goal is to take two halves of an array / vector and
 * sort these sections. These sections will be recursively divided until there is only one element left or the 
 * structure is empty, with the intent of merging them at the end.
 *
 * The merge function below is the meat of this algortihm. Essentially, this performs the array division.
 * The MergeSort function just recursively invokes merge to compile the final, sorted structure.
 *
 * The benefit of using a merge sort is that the complexity will usually be O(n log(n)) because it's always going to keep dividing sections. Since merge sorted elements are copied into another vector, the space complexity = O(n)
 *
 * MergeSort is probably the best basic sorting algorithm to use with Linked lists, if you have to sort them
 */
void merge(std::vector<int> &v, int left, int mid, int right) {
    std::vector<int> tmp;
    int firstIdxPtr = left;
    int secondIdxPtr = mid+1;

    while(firstIdxPtr <= mid && secondIdxPtr <= right) {
        if(v[firstIdxPtr] <= v[secondIdxPtr]) {
            tmp.push_back(v[firstIdxPtr]);
            firstIdxPtr++;
        } else {
            tmp.push_back(v[secondIdxPtr]);
            secondIdxPtr++;
        }
    }
    while(firstIdxPtr <= mid) {
        tmp.push_back(v[firstIdxPtr]);
        firstIdxPtr++;
    }
    while(secondIdxPtr <= right) {
        tmp.push_back(v[secondIdxPtr]);
        secondIdxPtr++;
    }
    for(int i = left; i <= right; i++) {
        v[i] = tmp[i-left];
    }
}


void mergeSort(std::vector<int> &v, int start, int end) {
    if(start >= end) return;
    int mid = start + (end - start) / 2;
    mergeSort(v, start, mid);
    mergeSort(v, mid+1, end);
    merge(v, start, mid, end);
}

int main() {
    std::vector<int> v7 = {6,5,4,3,2,1,0,-12};
    mergeSort(v7, 0, v7.size()-1);
    std::cout<<"MergeSort: ";
    print_array(v7);
}
